### PR TITLE
adds AllInOneTemplate Import feature

### PR DESCRIPTION
**Pull request by ruegenbfr**
- introduces upload data menue
- adds lib exceljs
- updates jest config: adds module uuid to transformIgnorePatterns
- refactors io.effects messages partially

ticket: #774

### CODE REVIEW

File: src/app/tracing/io/data-importer/xlsx-import/shared.ts [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+            }
+        }
+    });
+    return columnMappings;
+}
+
+export function getCleanedStringOrUndefined(
+    value: CellValue | undefined,
+): string | undefined {
+    if (typeof value === "string") {
```
Comment (j-hha) [inline src/app/tracing/io/data-importer/xlsx-import/shared.ts:179]: You want to return the string, as long as it is not empty. In all other cases, you want to return undefined, correct? I'd go for early returns here, e.g. sth like this

```
if (typeof value !== 'string' || value.trim().length === 0) {
    return undefined;
}

return value.trim();
```

File: src/app/tracing/io/data-importer/xlsx-import/shared.ts [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+        }
+    } else if (value !== undefined) {
+        return `${value}`;
+    }
+    return value;
+}
+
+function getCleanedInput(value: CellValue | undefined): CellValue | undefined {
+    if (typeof value === "string") {
+        value = value.trim();
```
Comment (j-hha) [inline src/app/tracing/io/data-importer/xlsx-import/shared.ts:192]: see above

File: src/app/tracing/io/data-importer/xlsx-import/shared.ts [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+    reqType: X,
+    addIssueCb: AddIssueCallback,
+    required = false,
+    invalidateRow = false,
+): TypeString2Type<X> | undefined {
+    const inputValue = getCleanedInput(row[colIndex]);
+    let value: TypeString2Type<X> | undefined;
+    if (inputValue === undefined) {
+        if (required === true) {
+            addIssueCb(
```
Comment (j-hha) [inline src/app/tracing/io/data-importer/xlsx-import/shared.ts:271]: same here: if there is a way to simplify the nested conditionals, I'd recommend that we do that

**Review by j-hha — COMMENTED** [review]
Preliminary review, more to come soon. First impression: Some of the nested conditionals could be simplyfied. I'd recommend that as it will make it easier to read and maintain the code.

File: src/app/tracing/io/data-importer/xlsx-import/shared.ts (hunk 1 above)
Comment (ruegenbfr) [reply src/app/tracing/io/data-importer/xlsx-import/shared.ts:178]: Done.

File: src/app/tracing/io/data-importer/xlsx-import/shared.ts (hunk 3 above)
Comment (ruegenbfr) [reply src/app/tracing/io/data-importer/xlsx-import/shared.ts:263]: I do not know how to unnest this without double checking things.

File: src/app/main-page/container/toolbar-action-container/toolbar-action-container.component.ts [hunk 4]
```diff
@@ -102,9 +102,15 @@ export class ToolbarActionContainerComponent implements OnInit, OnDestroy {
         this.loadFile(fileList);
     }
 
-    onSelectModelFile() {
+    // onSelectModelFile() {
```
[the remark this answers is not in the document]
Comment (ruegenbfr) [reply src/app/main-page/container/toolbar-action-container/toolbar-action-container.component.ts]: Yes. I removed it.

File: src/app/tracing/io/data-importer/xlsx-import/shared.ts (hunk 2 above)
Comment (ruegenbfr) [reply src/app/tracing/io/data-importer/xlsx-import/shared.ts]: Done.

File: src/app/main-page/presentation/toolbar-action/toolbar-action.component.html [hunk 5]
```diff
@@ -93,12 +93,21 @@
...
         mat-button
         data-cy="fcl-upload-model-button"
-        (click)="onSelectModelFile()"
+        [matMenuTriggerFor]="uploadMenu"
         matTooltip="Upload data"
     >
         <mat-icon>file_upload</mat-icon>
         Upload Data
     </button>
+    <!-- (click)="onSelectModelFile()" -->
```
Comment (arne-z) [inline src/app/main-page/presentation/toolbar-action/toolbar-action.component.html:102]: Is this comment still needed?

File: src/app/tracing/io/io.effects.ts [hunk 6]
```diff
@@ -34,6 +35,11 @@ import {
     MatLegacyDialogRef as MatDialogRef,
 } from "@angular/material/legacy-dialog";
 import { DataService } from "../services/data.service";
+import { ERROR_TEXTS, ERROR_RESOLUTION_TEXTS } from "./consts";
+
+function joinNonEmptyTexts(texts: string[], sep: string = " "): string {
+    return texts.filter((t) => t !== "").join(sep);
+}
```
Comment (arne-z) [inline src/app/tracing/io/io.effects.ts:42]: Should this live in non-ui-utils?

File: src/app/tracing/io/data-importer/xlsx-import/xlsx-reader.ts [hunk 7]
```diff
@@ -287,97 +295,160 @@ export interface Table {
...
+                    columnIndex,
+                )
+            ) {
                 return false;
             }
+
             if (typeof columnHeader !== "string") {
                 const childHeaders = columnHeader[1];
-                colSpan = getColumnHeadersSpan(childHeaders).colSpan;
+                // columnSpan = getColumnHeadersSpan(childHeaders).colSpan;
```
Comment (j-hha) [inline src/app/tracing/io/data-importer/xlsx-import/xlsx-reader.ts:414]: inentionally commented?

Comment (ruegenbfr) [reply src/app/tracing/io/data-importer/xlsx-import/xlsx-reader.ts:414]: no, I will remove it

File: src/app/main-page/presentation/toolbar-action/toolbar-action.component.html (hunk 5 above)
Comment (arne-z) [reply src/app/main-page/presentation/toolbar-action/toolbar-action.component.html:102]: Probably remove this comment?

**Review by arne-z — APPROVED** [review]

**Comment by arne-z** [comment]
See comment above about removing the comment in line 102 of 
`src/app/main-page/presentation/toolbar-action/toolbar-action.component.html` though that's not critical.

Other than that, looks good to me!

**Review by j-hha — APPROVED** [review]

**Comment by ruegenbfr** [comment]
I do not fix the comment in in line 102 of
`src/app/main-page/presentation/toolbar-action/toolbar-action.component.html` because hanna is working on this component and I don't that to generate unnecessary rebase conflicts.

**Comment by ruegenbfr** [comment]
The `// columnSpan...` comment was removed
Commits were squashed 
Squashed commit was rebased and merged.

**Pull request closed without merging**